Allows creating cubes from container objects programmatically

rivasdiaz opened the pull request:
#### Short description of what this resolves:

Allows creating cubes from container objects programmatically

#### Changes proposed in this pull request:

- A new interface injectable inside Arquillian test classes: ```ContainerObjectFactory```
- support for programmatically creating container objects with ```factory.createContainerObject(...)```. Created container objects will work exactly as if they were created by injecting them as fields. Associated Cube instances can also be stopped/destroyed using an injected ```CubeController```

**Fixes**: #564

rivasdiaz commented:
Some implementation details of this pull request:

It contains a big refactoring of ```CubeContainerObjectTestEnricher``` class. That class was originally in charge of:
- extracting information from the container object class
- extracting information from the field that would be initialized with the container object
- instantiating the container object
- creating/starting the associated ```Cube```.
- calling other ```TestEnricher``` instances on the created object and some built-in support for other annotations (```@CubeIp``` and ```@HostPort```)

After refactoring, everything related to the container object is handled by ```DockerContainerObjectBuilder``` class:
- extracting information from the container object class
- instantiating the container object
- creating/starting the associated ```Cube```.
- calling any registered ```TestEnricher``` instances (including ```CubeContainerObjectTestEnricher```) on the created object and some built-in support for other annotations (```@CubeIp``` and ```@HostPort```)

The class ```CubeContainerObjectTestEnricher``` now uses ```DockerContainerObjectFactory``` to create container objects. It still handles metadata extraction from fields annotated with ```@Cube``` and passes that information to the container object factory.

Creating container objects with the introduced ```DockerContainerObjectFactory``` is more powerful than ```CubeContainerObjectTestEnricher``` as it can receive any configuration. It enables, for example, the use of any ```Await``` strategy, something not currently supported with annotations.

All test cases that existed for ```CubeContainerObjectTestEnricher``` now depend on an injected instance of ```DockerContainerObjectFactory``` that does half of the work.

There is a new unit test class for ```DockerContainerObjectBuilder``` that contains an equivalent test case for each of the existing test cases in ```CubeContainerObjectTestEnricher```. I think that eventually the existing tests in ```CubeContainerObjectTestEnricher``` should be removed and simpler tests added, as the difference among test cases is now handled outside of the class being tested.

There are 2 more test cases in ```DockerContainerObjectBuilder``` for the built-in support for ```@CubeIp``` and ```@HostPort```. Those didn't exist in ```CubeContainerObjectTestEnricher```.

rivasdiaz commented:
cc @lordofthejars

bartoszmajsak commented:
Awesome stuff! Thanks a lot @rivasdiaz

lordofthejars commented on docker/docker/src/test/java/org/arquillian/cube/docker/impl/client/containerobject/CubeContainerObjectTestEnricherTest.java:
```diff
@@ -13,10 +12,14 @@
 import java.util.Collection;
+import java.util.Collections;
+
+import com.google.common.base.StandardSystemProperty;
```
I would not suggest to use Guava here. The main idea is that this dependency is transitive, this means that maybe in the future is not bundled anymore because a third library just removes its usage.

Also I would not add it as "real" dependency on project because it is a 3MB library for just using one class.

lordofthejars commented on docker/docker/src/test/java/org/arquillian/cube/docker/impl/client/containerobject/DockerContainerObjectBuilderTest.java:
```diff
@@ -0,0 +1,499 @@
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.*;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.*;
```
We tend to avoid start operator, I think that some IDEs do this automatically. No strong opinion here, so if you change ok if not is ok too.

lordofthejars commented on docker/docker/src/test/java/org/arquillian/cube/docker/impl/client/containerobject/DockerContainerObjectBuilderTest.java:
```diff
@@ -0,0 +1,499 @@
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import com.google.common.base.StandardSystemProperty;
```
Same problem related before CubeContainerObjectTestEnricherTest.java about Guava.

lordofthejars commented:
Also if you can update docs with this new enrichment (in Container Object section) would be awesome

rivasdiaz commented:
Thanks for the review @lordofthejars. I will be really busy the next three days. I'll try to work on your comments as soon as possible and update the pull request.

rivasdiaz commented:
cc @lordofthejars 
I finally completed all requested changes.

While reviewing the current documentation for Container Objects, I decided to change some of the wording on other sections to clarify its meaning. Hopefully as this PR touches a lot of areas related to container objects, those changes are OK too. Please let me know if you prefer me to only make a small change to the doc related to this PR, and I will rollback my other changes.

Also let me know if you have any other comments with the PR and I will do my best to try to fix them.

lordofthejars commented on api/src/main/java/org/arquillian/cube/ContainerObjectFactory.java:
```diff
@@ -0,0 +1,32 @@
+     * class. Some configuration can be overridden by passing an additional ContainerObjectConfiguration instance.
+     *
+     * @param containerObjectClass type of the container object to instantiate. Required
+     * @param configuration if not null, allows specifying some configuration parameters. Optional (can be null)
```
I have noticed this now, I don't like that users can add `null` value in public API. Of course there are some exceptions, but I usually prefer in public APIs to create a `Nullified` object or just create a new method. So for example:

`<T> T createContainerObject(Class<T> containerObjectClass)` and then if you want internally you call `this.createContainerObject(containerObjectClass, null)`

Another approach might be the nullified object which `ContainerObjectConfiguration` class has an static method called `defaults()` which returns an empty object and then inside the method again you can do something like:

```
if (configuration == ContainerObjectConfiguration.defaults()) {
  this.createContainerObject(containerObjectClass, null)
}
```

What I am trying to do is that the API does not expose the null in public API.

lordofthejars commented:
Final thing that I always miss as well, update docs after a change in the API :) but apart from this we are ready :)

rivasdiaz commented:
cc @lordofthejars 
Added a new method on the interface for clients to call if they don't want to pass any configuration.
Internally it passes an empty configuration to the overloaded method that expects a configuration. I also added a few validations that expect non null parameters.

lordofthejars commented:
Yes I have seen my point (and sorry to be a bit PITA) is to update this line https://github.com/arquillian/arquillian-cube/pull/574/files#diff-d2bcbb506a1f4067aa20b24e8c78f97cR254 to new method.

rivasdiaz commented:
I'm sorry for that. If that sample code were in a real test, it would have thrown an exception. The doc is fixed in my last commit. Unfortunately docs are not compiled :smile: 

Let me know if you see anything else. I will be happy to try to fix it

lordofthejars commented:
Nothing to forgive, this happens to best families. About compiling docs it is possible to do it with Asciidoctor and using `include` keyword, so you point directly to real test and you render docs in your compilation pipeline, so if there is a fail, then docs are not rendered. Sadly we cannot use this approach because of restrictions in Github.

Thank you very much for your help and your great work.

Alex.
